public class Solution {
    public List<List<Pair>> insertionSort(List<Pair> pairs) {
        List<List<Pair>> res = new ArrayList<>();

        for (int i = 0; i < pairs.size(); i++) {
            Pair cur = pairs.get(i);
            int j = i - 1;

            while (j >= 0 && pairs.get(j).key > cur.key) {
                pairs.set(j + 1, pairs.get(j));
                j--;
            }

            pairs.set(j + 1, cur);
            res.add(new ArrayList<>(pairs));
        }

        return res;
    }
}